### PR TITLE
[BC-breaking] Change dtype of softmax to support TorchScript and MyPy

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68584
* #68463
* #68461
* #68333
* __->__ #68336

Differential Revision: [D32470965](https://our.internmc.facebook.com/intern/diff/D32470965)

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Is this needed to land the masked softmax PR? This seems like a bigger change and might yield longer discussion. I'd suggest we reorder this stack even if the subsequent PR then has less features.

**Comment by pearu** [comment]
The intent of this PR was to fix https://github.com/pytorch/pytorch/runs/4211259175?check_suite_focus=true#step:6:25 , that is, a mypy failure in the masked softmax PR.

**Review by cpuhrsch — APPROVED** [review]
I'm going to accept and land this for now. I circulated this PR more widely and didn't get any objections. It also doesn't affect the public un-typed APIs. I'll tentatively mark this as BC breaking. Can you open an issue for this @pearu ? We'll also go ahead and land the subsequent, dependent PRs.

**Comment by cpuhrsch** [comment]
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32470965).

**Comment by pearu** [comment]
@cpuhrsch, created https://github.com/pytorch/pytorch/issues/68464 to reflect the possible BC-breaking change here.
